### PR TITLE
#164798915 User can follow and unfollow

### PR DESCRIPTION
**Pull request by geekdanitek**
#### Type of change
- [x]  Non-breaking change(introduces new endpoint to codebase)
#### How Has This Been Tested?
* All unit test passed
* Successfully followed a user using postman
Screenshot 2019-04-11 at 6 53 08 PM

* Successfully unfollowed a user using postman
Screenshot 2019-04-11 at 6 53 18 PM

#### Checklist:
- [x] Implement endpoint to follow a user
- [x] Implement endpoint to unfollow a user
- [x] Write unit test for `follow` user endpoint
- [x] Write unit test for `unfollow` user endpoint
- [x] Document `follow` and `unfollow` user endpoint
#### PIVOTAL BOARD
[#164798915](https://www.pivotaltracker.com/story/show/164798915)

### CODE REVIEW

File: server/controllers/profile.controller.js [hunk 1]
```diff
@@ -71,6 +71,68 @@ class ProfileController {
...
+    const newFollowing = await personToBeFollowed.addFollowers(requestingFollowership);
+    if (!newFollowing) {
+      return Response(res, 400, 'You cannot follow a user more than once');
+    }
+    const followers = await personToBeFollowed.getFollowers();
+    const userData = { personToBeFollowed: personToBeFollowed.username };
+    userData.followers = followers.map(follow => ({
+      follower: follow.username
+    }));
+    return Response(res, 200, successMessage, userData);
```
Comment (TheDrizzyWay) [inline server/controllers/profile.controller.js]: Very good job on the use of pivot tables and the many to many relationship. This just helped to clear up one aspect of sequelize I have been struggling with. However, your inline comment states that the above function returns the user that has been followed. Why then do we have an array of followers in the response? Do you think this information is useful to the user? Or did you add it for the sake of clarity?

File: server/controllers/profile.controller.js [hunk 2]
```diff
@@ -71,6 +71,68 @@ class ProfileController {
...
+     */
+  static async followUser(req, res) {
+    const { username } = req.params;
+    const { id } = req.user;
+    const requestingFollowership = await User.findByPk(id);
+    const personToBeFollowed = await User.findOne({ where: { username } });
+    const successMessage = 'Successfully followed user';
+
+    if (!personToBeFollowed) {
+      return Response(res, 404, 'You cannot follow a user that does not exist');
```
Comment (davidokonji) [inline server/controllers/profile.controller.js]: I like the way the error messages are very descriptive, good work

File: server/test/follow.spec.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    expect(res.body.message).to.equal('You cannot unfollow a user more than once');
+  });
+  it('should not unfollow a user if it is the same user requesting to be followed', async () => {
+    const res = await chai.request(app)
+      .delete('/api/v1/profiles/Mofe/follow')
+      .set({ Authorization: `Bearer ${userToken}` });
+    expect(res).to.have.status(409);
+    expect(res.body.message).to.equal('You cannot unfollow yourself');
+  });
+});
```
Comment (davidokonji) [inline server/test/follow.spec.js:82]: Detailed test cases, good one

File: server/controllers/profile.controller.js (hunk 1 above)
Comment (geekdanitek) [reply server/controllers/profile.controller.js]: @TheDrizzyWay Yeah the response data containing the array of followers is to show the user has been added to another user followership list.

File: server/test/follow.spec.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    expect(res.body.data).to.have.property('followers');
+  });
+  it('should not follow a user if the user doesnt exist', async () => {
+    const res = await chai.request(app)
+      .post('/api/v1/profiles/Dimkpayy/follow')
+      .set({ Authorization: `Bearer ${userToken}` });
+    expect(res).to.have.status(404);
+    expect(res.body.message).to.equal('You cannot follow a user that does not exist');
+  });
+  it('should not a follow a user if the user has been followed before', async () => {
```
Comment (samorahno) [inline server/test/follow.spec.js]: Good job @geekdanitek, kindly check this line for grammatical errors

Comment (geekdanitek) [reply server/test/follow.spec.js]: Ok, eagles eyes 👍

File: server/controllers/profile.controller.js [hunk 5]
```diff
@@ -71,6 +71,68 @@ class ProfileController {
...
+  /**
+     * @description follows another user
+     * @param {object} req
+     * @param {object} res
+     * @returns {object} user that has been followed
+     */
+  static async followUser(req, res) {
+    const { username } = req.params;
+    const { id } = req.user;
+    const requestingFollowership = await User.findByPk(id);
```
Comment (paradisekelechi) [inline server/controllers/profile.controller.js]: `requestingFollowership` - Poor variable naming

File: server/controllers/profile.controller.js [hunk 6]
```diff
@@ -71,6 +71,68 @@ class ProfileController {
...
+    const personToBeFollowed = await User.findOne({ where: { username } });
```
Comment (paradisekelechi) [inline server/controllers/profile.controller.js]: `personToBeFollowed` - Poor variable naming. Use `follower` and `followee`. That makes more sense

File: server/controllers/profile.controller.js [hunk 7]
```diff
@@ -71,6 +71,68 @@ class ProfileController {
...
+    const successMessage = 'Successfully followed user';
```
Comment (paradisekelechi) [inline server/controllers/profile.controller.js]: Refactor this

File: server/controllers/profile.controller.js (hunk 2 above)
Comment (paradisekelechi) [inline server/controllers/profile.controller.js]: Keep the response simple and straight to the point. `User does not exist` could suffice

File: server/controllers/profile.controller.js [hunk 8]
```diff
@@ -71,6 +71,68 @@ class ProfileController {
...
+
+    if (!personToBeFollowed) {
+      return Response(res, 404, 'You cannot follow a user that does not exist');
+    }
+    if (requestingFollowership.id === personToBeFollowed.id) {
+      return Response(res, 409, 'You cannot follow yourself');
+    }
+    const newFollowing = await personToBeFollowed.addFollowers(requestingFollowership);
```
Comment (paradisekelechi) [inline server/controllers/profile.controller.js]: Poor variable naming

File: server/controllers/profile.controller.js [hunk 9]
```diff
@@ -71,6 +71,68 @@ class ProfileController {
...
+    if (!newFollowing) {
+      return Response(res, 400, 'You cannot follow a user more than once');
+    }
+    const followers = await personToBeFollowed.getFollowers();
+    const userData = { personToBeFollowed: personToBeFollowed.username };
+    userData.followers = followers.map(follow => ({
+      follower: follow.username
+    }));
```
Comment (paradisekelechi) [inline server/controllers/profile.controller.js]: Make this block of code better and more readable/understandable

File: server/controllers/profile.controller.js [hunk 10]
```diff
@@ -71,6 +71,68 @@ class ProfileController {
...
+  }
+
+  /**
+     * @description unfollows a user
+     * @param {object} req
+     * @param {object} res
+     * @returns {object} user that has been unfollowed
+     */
+
+  static async unFollowUser(req, res) {
```
Comment (paradisekelechi) [inline server/controllers/profile.controller.js]: `unFollowUser` does not go well. `unfollowUser` is a better naming

File: server/controllers/profile.controller.js [hunk 11]
```diff
@@ -71,6 +71,68 @@ class ProfileController {
...
+    const { username } = req.params;
+    const { id } = req.user;
+    const requestingUnFollowership = await User.findByPk(id);
```
Comment (paradisekelechi) [inline server/controllers/profile.controller.js]: Poor variable naming

File: server/controllers/profile.controller.js [hunk 12]
```diff
@@ -71,6 +71,68 @@ class ProfileController {
...
+    const personToBeUnFollowed = await User.findOne({ where: { username } });
```
Comment (paradisekelechi) [inline server/controllers/profile.controller.js]: Poor variable naming

File: server/controllers/profile.controller.js [hunk 13]
```diff
@@ -71,6 +71,68 @@ class ProfileController {
...
+    const successMessage = 'Successfully unfollowed user';
```
Comment (paradisekelechi) [inline server/controllers/profile.controller.js]: Refactor this

**Review by paradisekelechi — CHANGES_REQUESTED** [review]

File: server/controllers/profile.controller.js (hunk 6 above)
Comment (geekdanitek) [reply server/controllers/profile.controller.js]: Ok will implement

File: server/controllers/profile.controller.js [hunk 14]
```diff
@@ -71,6 +71,66 @@ class ProfileController {
...
+    }
+    if (follower.id === followee.id) {
+      return Response(res, 409, 'Cannot follow self');
+    }
+    const newFollower = await followee.addFollowers(follower);
+    if (!newFollower) {
+      return Response(res, 400, 'Cannot follow user twice');
+    }
+    const followers = await followee.getFollowers();
+    const user = {
```
Comment (davidokonji) [inline server/controllers/profile.controller.js:98]: Good work refactoring

**Review by davidokonji — APPROVED** [review]

**Review by samorahno — APPROVED** [review]

**Review by chukwudidu — APPROVED** [review]
Good work

**Review by TheDrizzyWay — APPROVED** [review]
Good to see that you have implemented the requested changes.

**Review by BukkyOmo — APPROVED** [review]

**Review by paradisekelechi — APPROVED** [review]

**Pull request merged**